src/inmanta/deploy/state.py

```diff
@@ -125,6 +181,73 @@ def reset(self) -> None:
+        blocked_status = BlockedStatus.TRANSIENT if transient else BlockedStatus.YES
+        if resource in self.resource_state:
+            if not transient:
+                self.resource_state[resource].status = ResourceStatus.UNDEFINED
+            if self.resource_state[resource].blocked is not BlockedStatus.YES:
+                self.resource_state[resource].blocked = blocked_status
+        else:
+            resource_status = ResourceStatus.UNDEFINED if not transient else ResourceStatus.HAS_UPDATE
+            self.resource_state[resource] = ResourceState(
+                status=resource_status, deployment_result=DeploymentResult.NEW, blocked=blocked_status
+            )
+            self.types_per_agent[details.id.agent_name][details.id.entity_type] += 1
+        self.dirty.discard(resource)
+
+    def block_provides(self, resources: set[ResourceIdStr]) -> set[ResourceIdStr]:
+        """
+        Marks the provides of the given resources as blocked.
+
+        Must be called under the scheduler lock. This method assumes that all the resources of the model version
+        are populated in the resources dictionary.
+
+        :param resources: The set of resources for which the provides have to be blocked.
+        :return: The set of dependent resources that were marked as blocked (transitively).
+        """
+        result = set()
+        for dependent_resource in self.requires.get_all_provides_transitively(resources):
+            self.block_resource(dependent_resource, self.resources[dependent_resource], transient=True)
+            result.add(dependent_resource)
+        return result
+
+    def unblock_resource(self, resource: ResourceIdStr) -> None:
+        """
+        Mark the given resource no longer as blocked. Also mark the provides of the given resource as unblocked (transitively)
+        if they become unblocked because of this.
+
+        Must be called under the scheduler lock. This method assumes that all the resources of the model version
+        are populated in the resources dictionary.
+
+        :param resource: The resource that has to be unblocked and potentially any of its provides as a consequence of
+                         `resource` being unblocked.
+        """
+
+        def _unblock(resource: ResourceIdStr) -> None:
+            self.resource_state[resource].blocked = BlockedStatus.NO
+            if self.resource_state[resource].status is ResourceStatus.UNDEFINED:
+                self.resource_state[resource].status = ResourceStatus.HAS_UPDATE
+            if self.resource_state[resource].status is ResourceStatus.HAS_UPDATE:
+                self.dirty.add(resource)
+
+        _unblock(resource)
+        for res in self.requires.get_all_provides_transitively(resource):
+            is_blocked = self.resource_state[res].status is ResourceStatus.UNDEFINED or any(
+                self.resource_state[r].blocked.is_blocked() for r in self.requires.get(res, set())
```

Comment:
The `get_all_provides_transitively()` method returns the element sorted. That's why it's a list instead of a set. But I didn't document this. I added the documentation now.